### PR TITLE
ENH: Use GLSL pipeline for efficient draws

### LINKED ISSUE #251: FIX: about 4% of flips are happening late
@lkishline and I are getting an AV experiment set up, which involves some carefully timed flashes. We are timing these flips such that there is a first flip when `start_stimulus` is called and all other flips use the `when` argument relative to that one. We have set it up such that the time we pass for `when` is equal to the initial flip time + a time corresponding to a whole number of frames - `0.9 / frame_rate`. This should very reliably put the flips where we want them.

Instead, we are losing about 10% of trials because some trials have late flips, sometimes by even more than one frame. @Eric89GXL @drammock, do you guys have any idea how the flip timing could be made more reliable?

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #251.

@drammock @rkmaddox @lkishline can you try this branch to see if you get better results?

### CODE REVIEW

**Comment by Eric89GXL** [comment]
Here is current `master` for me:
master

And on this branch:
glsl

Note the y-axes are different. So now 100 and 500 are both fast and reliable, and 500 takes ~400 microseconds to draw on my machine. Hopefully you see something similar.

**Comment by Eric89GXL** [comment]
I'll need to make some tweaks to Appveyor to get it to build and work properly there, but at least Travis is happy and tests pass locally for now.

**Comment by drammock** [comment]
Thanks @Eric89GXL. Will test tomorrow. since circles default to 200 edges
this should hopefully solve the current problem. I'm still puzzled that you
see a few long delays in the plot for small numbers of edges though... Any
speculations as to why? If you generate the plot again is it deterministic?

**Comment by rkmaddox** [comment]
Works very well on my computer. @lkishline will post test results from the lab computer imminently.

Thanks for the fix.

File: expyfun/_input_controllers.py [hunk 1]
```diff
@@ -421,15 +421,11 @@ def _point_in_object(self, pos, obj):
...
-            tri = np.array([[obj._points[2 * p], obj._points[2 * p + 1]] for
-                            p in points])
-            found = self._point_in_tri(pos, tri)
-            index += 3
-        return found
+        these_tris = obj._tris['fill'].reshape(-1, 3)
+        for tri in these_tris:
+            if self._point_in_tri(pos, obj._points['fill'][tri]):
+                return True
+        return False
```
Comment (rkmaddox) [inline expyfun/_input_controllers.py]: Nice change. Much cleaner.

**Comment by lkishline** [comment]
Results on MARLO, not so hot:

drawtest

**Comment by lkishline** [comment]
Here is the same graph, except with `audio_controller='pyglet', response_device='keyboard', trigger_controller='dummy'` passed to `ExperimentController`:

drawtest-pyglet

!!!!!

**Comment by rkmaddox** [comment]
So Lindsey and I ran it some more, and it basically works pretty darn well, regardless of the settings. I have no idea why you guys previously found a difference based on whether TDT was in use or not, but there no longer seems to be one. I have no ideas on what might be controlling it. @lkishline, did you ever run it on a bunch more blocks to see if it was still perfect?

**Comment by lkishline** [comment]
@rkmaddox was just about to run it on 4 blocks and got a frame rate error on our assertion line....not sure why...But tried to run it again and it started up just fine. It is currently in the middle of running 4 blocks.

**Comment by lkishline** [comment]
Alright, well I ran it on 4 blocks and only saw one missed frame!

Although one of the fixes made the fixation dot look different and text between my blocks disappear. But it could be cause I'm using a different version of expyfun then what I tested my script on.

**Comment by rkmaddox** [comment]
Good that there are far fewer missed frames. Losing 1 in 400 trials seems like quite acceptable attrition.

Don't know why the fixation dot would look different. Does the fixation dot with default parameters look right? On my linux box it looks normal using this branch.

**Comment by lkishline** [comment]
Oh wait - never mind. I was using the ConcentricCircles for the fixation dot and the answer cue dot because I was moving them around and changing the colors/ sizes.

**Comment by drammock** [comment]
@lkishline is the text issue similar to what is described in #182, or is the text just simply gone?

**Comment by lkishline** [comment]
@drammock I don't think so? screen_prompt text within in my block/trials loop looks like this. Just two dots in the middle of the screen. And when I hit the button it turns to one dot, and then I hit the button to continue again and it goes into the trial. The text before the block loop is fine.
image

**Comment by drammock** [comment]
that is definitely something different.

**Comment by lkishline** [comment]
Yes, I thought I would take it upon myself to create as many mysterious issues as possible. So far I think I'm doing pretty well on that endeavor.

**Comment by Eric89GXL** [comment]
Can you share a minimal example recreating the behavior?

**Comment by lkishline** [comment]
@Eric89GXL Yes, I will give that a go. Give me a bit.

**Comment by lkishline** [comment]
Okay, so hopefully all this gist stuff works....[gist] (https://gist.github.com/lkishline/3d7f35b0ca428caadc3f)

**Comment by Eric89GXL** [comment]
Do there need to be multiple blocks to see the problem? Multiple trials? In other words, is that example really minimal to reproduce the issue?

**Comment by lkishline** [comment]
@Eric89GXL ah, right. let me go back and test.

**Comment by Eric89GXL** [comment]
Cool -- the "minimal" idea is important because it makes a pretty good step toward isolating the problem to a small set of calls, which is why I'm being picky about it.

**Comment by lkishline** [comment]
Yea, I just realized that by not making it truly minimal I was asking others to do more work. My bad!

**Comment by lkishline** [comment]
[Updated!] (https://gist.github.com/lkishline/3d7f35b0ca428caadc3f)

**Comment by drammock** [comment]
@lkishline's second gist works normally for me on Linux (moctezuma).  Will try to look at it on marlo tomorrow morning.

**Comment by Eric89GXL** [comment]
I can replicate on Windows, fix coming

**Comment by Eric89GXL** [comment]
@lkishline updated, example works over here now. Feel free to give it a try. FYI this is the example after a couple other minor tweaks to facilitate testing:
```
from expyfun import ExperimentController
from expyfun.visual import ConcentricCircles

with ExperimentController('Test', full_screen=False, participant='foo',
                          session='bar', output_dir=None) as ec:
    ec.screen_prompt('a', max_wait=0.3)
    dot = ConcentricCircles(ec)
    dot.draw()
    ec.flip()
    ec.wait_secs(0.3)
    ec.screen_prompt('b', max_wait=0.3)
```

**Comment by lkishline** [comment]
@Eric89GXL It works! Thanks so much. 

And I'll work on creating better minimal examples in the future.

**Comment by drammock** [comment]
I'm guessing @rkmaddox is the one of us best-qualified to vet your OpenGL code, @Eric89GXL, so I'll leave the merge decision to him.  Otherwise LGTM.

**Comment by rkmaddox** [comment]
Yikes. I have zero GL experience. Will do my best, though.

**Comment by Eric89GXL** [comment]
In case you want to learn a bit of background:

http://duriansoftware.com/joe/An-intro-to-modern-OpenGL.-Chapter-1:-The-Graphics-Pipeline.html

File: expyfun/visual/_visual.py [hunk 2]
```diff
@@ -176,27 +306,39 @@ def set_line_width(self, line_width):
...
+                                self._buffers[kind]['array'])
+                loc_pos = gl.glGetAttribLocation(self._program, b'a_position')
+                gl.glEnableVertexAttribArray(loc_pos)
+                gl.glVertexAttribPointer(loc_pos, 2, gl.GL_FLOAT, gl.GL_FALSE,
+                                         0, 0)
+                loc_col = gl.glGetUniformLocation(self._program, b'u_color')
+                gl.glUniform4f(loc_col, *self._colors[kind])
+                cmd()
+                # The following line is probably only necessary because
+                # Pyglet makes some ossumptions about the GL state that
```
Comment (rkmaddox) [inline expyfun/visual/_visual.py]: typo (assumptions)

**Comment by rkmaddox** [comment]
Other than a single typo in the comments, I think it looks good (but admittedly, much of it was not understandable to me). I think we should watch our stimuli carefully as we develop the next couple experiments to make sure things are really working. And hey, maybe this will fix the text spacing bug?

+1 for merge.

**Comment by Eric89GXL** [comment]
This doesn't touch TextVisual at all unfortunately, so it shouldn't affect that bug. I still saw it while debugging :(

Typo fixed.

**Comment by rkmaddox** [comment]
Thanks for the vast improvement.

**Pull request merged**